src/backend.f90

```diff
@@ -0,0 +1,179 @@
+      !! at run time and therefore backend implementations are responsible from
+      !! assigning these pointers to the right subroutines.
+      !!
+      !! trans_?2? subroutines are straightforward, they rearrange data into
+      !! our specialist data structure so that regardless of the direction
+      !! tridiagonal systems are solved efficiently and fast.
+      !!
+      !! Allocator is responsible from giving us a field sized array when
+      !! requested. For example, when the derivations in x direction are
+      !! completed and we are ready for the y directional derivatives, we need
+      !! three fields to reorder and store the velocities in y direction. Also,
+      !! we need three more fields for storing the results, and the get_block
+      !! method of the allocator is used to arrange all these memory
+      !! assignments. Later, when a field is no more required, release_block
+      !! method of the allocator can be used to make this field available
+      !! for later use.
+      real :: nu
+      class(allocator_t), pointer :: allocator
+      class(dirps_t), pointer :: xdirps, ydirps, zdirps
+   contains
+      procedure :: transeq
+      procedure :: run
+      procedure(transeq_ders), deferred :: transeq_x
+      procedure(transeq_ders), deferred :: transeq_y
+      procedure(transeq_ders), deferred :: transeq_z
+      procedure(transposer), deferred :: trans_x2y
+      procedure(transposer), deferred :: trans_x2z
+      procedure(sum9into3), deferred :: sum_yzintox
+   end type base_backend_t
+
+   abstract interface
+      subroutine transeq_ders(self, du, duu, d2u, u, v, w, dirps)
+         import :: base_backend_t
+         import :: field_t
+         import :: dirps_t
+         implicit none
+
+         class(base_backend_t) :: self
+         class(field_t), intent(out) :: du, duu, d2u
+         class(field_t), intent(in) :: u, v, w
+         type(dirps_t), intent(in) :: dirps
+      end subroutine transeq_ders
+   end interface
+
+   abstract interface
+      subroutine transposer(self, u_, v_, w_, u, v, w)
+         import :: base_backend_t
+         import :: field_t
+         implicit none
+
+         class(base_backend_t) :: self
+         class(field_t), intent(out) :: u_, v_, w_
+         class(field_t), intent(in) :: u, v, w
+      end subroutine transposer
+   end interface
+
+   abstract interface
+      subroutine sum9into3(self, du, dv, dw, du_y, dv_y, dw_y, du_z, dv_z, dw_z)
+         import :: base_backend_t
+         import :: field_t
+         implicit none
+
+         class(base_backend_t) :: self
+         class(field_t), intent(inout) :: du, dv, dw
+         class(field_t), intent(in) :: du_y, dv_y, dw_y, du_z, dv_z, dw_z
+      end subroutine sum9into3
+   end interface
+
+contains
+
+   subroutine transeq(self, du, dv, dw, u, v, w)
+      implicit none
+
+      class(base_backend_t) :: self
+      class(field_t), intent(out) :: du, dv, dw
```

Comment:
I think that technically this will create new fields, deallocating their `allocatable` members (which likely get reallocated later when set).

Reply:
I use intent(out) in the kernels as well and haven't really noticed a deallocation/allocation and an associated performance loss due to this. Still worth double checking with using intent(inout) instead, I'll let you know.

Reply:
Actually, do you expect such a deallocation/allocation when you pass an allocatable array into a subroutine and within the subroutine declare intent(out) for the dummy variable corresponding to the allocatable array?

Reply:
or this only applies to user defined types?

Reply:
It is complicated - if the _argument_ is given the attribute ``allocatable`` and ``intent(out)`` it will be deallocated on entry. An allocatable array that is passed otherwise will not be deallocated, e.g.

```
subroutine foo(a, b)
  integer, dimension(:), allocatable, intent(out) :: a ! Deallocated on entry
  integer, dimension(:), intent(out) :: b ! Not deallocated
  ...
```

For user defined types I think they will technically be reinitialised with ``intent(out)`` so if they contain an ``allocatable`` field it will be deallocated.

Reply:
Alright then we're fine for the kernels because I never use allocatable attribute for the kernel subroutines. But for the user defined types it is better to use intent(inout) to avoid reallocation. I didn't know about this at all, thanks!